### PR TITLE
[SPARK-47912][SQL] Infer serde class from format classes

### PR DESCRIPTION
**Pull request by wForget**
### What changes were proposed in this pull request?

Infer serde class from format classes.

### Why are the changes needed?

File format of insert overwrite dir does not take effect.

Reproduce:
```
set hive.default.fileformat=parquet;

-- Not as expected, I want an orc file but generated a parquet file.
INSERT OVERWRITE DIRECTORY '$tablePath'
STORED AS
  INPUTFORMAT 'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'
  OUTPUTFORMAT 'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat'
select 1;

-- Failed
select * from `orc`.`$tablePath`; 
```

When we use stored as fileformat to insert directory, `InsertIntoDir.storage.serde` will be set to `parquet` which is default fileformat serde class we set, and then `RelationConversions` optimizer will convert `InsertIntoDir` to `InsertIntoDataSourceDirCommand` causing the fileformat classes we specified to be ignored.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

added new unit test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by wForget** [comment]
@cloud-fan could you please take a look?

**Comment by smileyboy2019** [comment]
Structured Streaming supports writing Spark SQL and using SQL to write stream processing logic. Is this possible, similar to the syntax of Flink. SQL can satisfy the flow processing process.

**Comment by wForget** [comment]
> Structured Streaming supports writing Spark SQL and using SQL to write stream processing logic. Is this possible, similar to the syntax of Flink. SQL can satisfy the flow processing process.

This doesn't seem to be relevant to the current PR, you can get more help from https://spark.apache.org/community.html.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala [hunk 1]
```diff
@@ -96,6 +96,16 @@ object HiveSerDe {
     serdeMap.get(key)
   }
 
+  /**
+   * Get serde class from format classes
+   */
+  def getSerDeFromFormatClasses(inputFormat: String, outputFormat: String): Option[String] = {
+    serdeMap.find {
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/HiveSerDe.scala:103]: If the in/out mismatch, is it suitable to fall back to LazySimpleSerDe?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -681,11 +681,14 @@ class SparkSqlAstBuilder extends AstBuilder {
     } else {
       val serdeInfo = maybeSerdeInfo.get
       if (serdeInfo.storedAs.isEmpty) {
+        val serde = serdeInfo.formatClasses
+          .flatMap(f => HiveSerDe.getSerDeFromFormatClasses(f.input, f.output))
+          .orElse(serdeInfo.serde)
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:686]: What are the expected behaviors of both FormatClasses and ROW FORMAT SERDE are given but conflict

Comment (wForget) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Thanks for your review. I have made some adjustments to prioritize the user-defined 'ROW FORMAT SERDE'.

**Comment by wForget** [comment]
Thanks @yaooqinn , I rethought this PR and it seems unreasonable. I will avoid this issue by adding `spark.sql.hive.convertMetastoreInsertDir=false;`.

**Pull request closed without merging**